assets/model_monitoring/components/src/token_statistics_compute_metrics/compute_token_statistics_metrics.py

```diff
@@ -0,0 +1,351 @@
+        print(f"Error: The number of rows in the numerator_metric_df is greater the number of rows in the denominator_metric_df")
+        print(f"numerator_metric_df count: {numerator_metric_df.count()}")
+        print(f"numerator_metric_df sample: {numerator_metric_df.take(10)}")
+        print(f"denominator_metric_df count: {denominator_metric_df.count()}")
+        print(f"denominator_metric_df sample: {denominator_metric_df.take(10)}")
+        return spark.createDataFrame([], numerator_metric_df.schema)
+
+    ratio_df = numerator_metric_df.withColumnRenamed("metric_value","numerator")\
+    .join(
+            denominator_metric_df.withColumnRenamed("metric_value","denominator"), on=dimensions, how="rightouter",
+        ).select(dimensions+["numerator","denominator"])
+
+    # if the numerator has null values then replace them with 0 
+    ratio_df = ratio_df.fillna(0, subset=["numerator"])
+
+    ratio_df = ratio_df.withColumn(
+            "metric_value",
+            ratio_df["numerator"]
+            / ratio_df["denominator"]
+            * 100,
+        )
+    ratio_df = ratio_df.withColumn(
+            "metric_name", lit(ratio_metric_name)
+        )
+    ratio_df = ratio_df.select(
+            dimensions + ["metric_name", "metric_value"]
+        )
+    
+    # Check that the number of rows in the ratio_df is same as the number of rows in the denominator_metric_df
+    # Else something wrong likely happened in the join
+    if ratio_df.count() != denominator_metric_df.count():
+        print(f"Error: The number of rows in the ratio_df is not same as the number of rows in the numerator_metric_df")
+        print(f"denominator_metric_df count: {denominator_metric_df.count()}")
+        print(f"denominator_metric_df sample: {denominator_metric_df.take(10)}")
+        print(f"ratio_df count: {ratio_df.count()}")
+        print(f"ratio_df sample: {ratio_df.take(10)}")
+        return
+    
+    return ratio_df
+
+def compute_GPU_utilization_metrics(token_df):
+    """Compute GPU utilization metrics for each group and group_pivot:
+    - Expect the token_df to have the following columns:group, group_pivot
+    - Calls failed due to overload: 
+        -- total number of calls with 429 status code
+        -- %age of calls with 429 status code
+    - Calls Succeeded:
+        -- total number of calls with 200 status code
+        -- %age of calls with 200 status code
+    - GPU utlization for the successful calls
+        -- Sum of prompt_tokens, completion_tokens, total_tokens for calls with 200 status code
+        -- Average of prompt_tokens, completion_tokens, total_tokens for calls with 200 status code
+    """
+    # check if the token_df has the expected dimension columns
+    dimensions=['group','group_pivot']
+    if set(dimensions).issubset(token_df.columns) == False:
+        return
+
+    # Call Counts
+    call_counts = token_df.groupBy(dimensions).agg(
+    lit("num_calls").alias('metric_name'),
```

Comment:
nit: a level of indentation will be nice.

Reply:
Thanks for catching. done.